api/v1alpha1/bundle.go

```diff
@@ -46,7 +49,8 @@ import "strings"
 	instances: [string & =~"^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$" & strings.MaxRunes(63) & strings.MinRunes(1)]: {
 		module: close({
 			url:     string & =~"^oci://.*$"
-			version: string & strings.MinRunes(3)
+			version: *"latest" | string
```

Comment:
I'm not sure how defaults are applied in Cue; I suspect that if you run a value against this schema, then you get the unified version out the other end, and that will have the default applied. That means there's no distinguishing between these cases:

| version | digest | meaning |
|-----------|----------|-------------|
| `""`       | &lt;sha256&gt; | use the digest to fetch the image |
| `"latest"` | &lt;sha256&gt; | fetch the image tagged "latest", check that it has the digest given |

As it happens, the code here https://github.com/stefanprodan/timoni/pull/48/files#diff-54c697a1927e20930729129477a32e149e0285d47c7c05d2e60f72ad8395d595R144 resolves this by treating both `""` and `latest` as meaning " pull the latest, whatever the digest says". This is a bit confusing, but probably OK if it's explicitly called out in help docs.

Could it be better? If you didn't supply a default in the schema, you could distinguish between those cases, and distinguish between "pull the latest and check it has this digest", and "pull the latest, whatever it is". I can't see much use for that first one, but it is different to "just pull latest, whatever it is" -- perhaps you'd use it if you have automation to update the digest, and didn't want apply to proceed if that hasn't happened yet.

Reply:
> As it happens, the code here https://github.com/stefanprodan/timoni/pull/48/files#diff-54c697a1927e20930729129477a32e149e0285d47c7c05d2e60f72ad8395d595R144 resolves this by treating both "" (to which the default is applied) and latest (supplied explicitly) as meaning " pull the latest, whatever the digest says".

This is not the case, if the digest is specified and version is empty/latest, it will use the digest and not look at the latest digest. The `latest` value has a special meaning to Timoni as it does to Docker, and is ignored when coupled with digest. The digest check is only performed when semver version is set.

Reply:
I've added a test to prove this, "creates instance for digest ignoring latest".

Reply:
Owps, yes I got the explanation of that code the wrong way around, sorry! It's still a special case, but now I understand what it's about -- docker pull will treat e.g., `docker pull nginx:latest@sha256:<some other SHA> as though `latest` wasn't there.